src/dma/mod.rs

```diff
@@ -0,0 +1,1171 @@
+        }
+        self.stream.clear_transfer_complete_interrupt();
+
+        if self.double_buf.is_some() && DIR::direction() != DmaDirection::MemoryToMemory {
+            // NOTE(unsafe) We now own this buffer and we won't call any &mut methods on `new` until
+            // the end of the DMA transfer
+            let (new_buf_ptr, new_buf_len) = unsafe { new_buf.write_buffer() };
+
+            // We can't change the transfer length while double buffering
+            if new_buf_len < usize::from(self.transfer_length) {
+                return Err(DMAError::SmallBuffer(new_buf));
+            }
+
+            if STREAM::current_buffer() == CurrentBuffer::DoubleBuffer {
+                self.stream.set_memory_address(new_buf_ptr as u32);
+                let old_buf = self.buf.replace(new_buf);
+
+                // "Subsequent reads and writes cannot be moved ahead of preceding reads"
+                compiler_fence(Ordering::Acquire);
+
+                // We always have a buffer, so unwrap can't fail
+                return Ok((old_buf.unwrap(), CurrentBuffer::FirstBuffer));
+            } else {
+                self.stream
+                    .set_memory_double_buffer_address(new_buf_ptr as u32);
+                let old_buf = self.double_buf.replace(new_buf);
+
+                // "Subsequent reads and writes cannot be moved ahead of preceding reads"
+                compiler_fence(Ordering::Acquire);
+
+                // double buffering, unwrap can never fail
+                return Ok((old_buf.unwrap(), CurrentBuffer::DoubleBuffer));
+            }
+        }
+        self.stream.disable();
+
+        // "No re-ordering of reads and writes across this point is allowed"
+        compiler_fence(Ordering::SeqCst);
+
+        // NOTE(unsafe) We now own this buffer and we won't call any &mut methods on `new` until
+        // the end of the DMA transfer
+        let (buf_ptr, buf_len) = unsafe { new_buf.write_buffer() };
+        self.stream.set_memory_address(buf_ptr as u32);
+        self.stream.set_number_of_transfers(buf_len as u16);
+        let old_buf = self.buf.replace(new_buf);
+
+        // "Preceding reads and writes cannot be moved past subsequent writes"
+        compiler_fence(Ordering::Release);
+
+        unsafe {
+            self.stream.enable();
+        }
+
+        Ok((old_buf.unwrap(), CurrentBuffer::FirstBuffer))
+    }
+
+    /// Stops the stream and returns the underlying resources.
+    pub fn free(mut self) -> (STREAM, PERIPHERAL, BUF, Option<BUF>) {
+        self.stream.disable();
+        compiler_fence(Ordering::SeqCst);
+        self.stream.clear_interrupts();
+
+        unsafe {
+            let stream = ptr::read(&self.stream);
+            let peripheral = ptr::read(&self.peripheral);
+            let buf = ptr::read(&self.buf);
+            let double_buf = ptr::read(&self.double_buf);
+            mem::forget(self);
+            (stream, peripheral, buf.unwrap(), double_buf)
+        }
```

Comment:
> Is that done because we cannot allow Drop to run because it would access the already moved out stream?

Yes, the compiler doesn't let you destruct a type that implements `Drop`. It would give a compiler error.

> The Option approach used in the Embedonomicon might be more inconvenient to implement, but it doesn't have this drawback.

The `Option` approach has the `unwrap` operation every time you want to access the inner transfer, which is a considerable drawback IMO (performance and convenience wise), and the `ptr::read` seems to be way more likely to be optimized by the compiler, especially given that both `peripheral` and `stream` are zero sized types and the buffer is a pointer type which probably is quite small.

Reply:
`Option` has the drawback of being inconvenient, but only for HAL writers. Users of the API don't have to notice it. And I'm pretty sure you can define an internal helper method to make it easier inside the HAL too. Regarding performance, I don't think it would be too bad, but if you are concerned about that there are `unsafe` ways to tell the compiler that the option is always `Some` (outside of the `Drop` impl), so performance should stay the same.

But I agree that using an `Option` is a hassle so we should only use it if your approach has real drawbacks. What concerns me is that we leak memory here and I'm not sure that gets optimized away. If it doesn't you leave the user with no choice than to leak memory if they want to get their resources back. If you have an application that does continuous one-shot transfers, at some point all the available memory will be leaked, which would make the API unusable.

Reply:
I'm not sure what you think will leak here, the only types that we don't "move out" with `ptr::read` are `PhantomData` and an `u16` both of which don't implement (and never will) `Drop`.

Reply:
I think my misconception was that when I saw `mem::forget` I though "leak". But in this case we only `forget` stack-allocated memory, which will be freed when that stack frame is popped, so there is no actual leak here.

So I think you are right and this is okay to do. I'd still very much like a comment describing why it is done that way. And also why it doesn't leak (so people like me don't get confused).